distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso

```diff
@@ -599,12 +599,41 @@ Text.is_empty = this == ""
 Text.not_empty : Boolean
 Text.not_empty = this.is_empty.not
 
+## Inserts text value at the specified index.
+
+   Arguments:
+     - index: The location in the text to inset text at (0-based). The
+       index is also allowed to be negative, then the characters are
+       counted from the end of the text, i.e. -1 will insert the text after
+       the last x.
+
+   ! What is a Character?
+     A character is defined as an Extended Grapheme Cluster, see Unicode
+     Standard Annex 29. This is the smallest unit that still has semantic
+     meaning in most text-processing applications.
+
+   > Example
+    Insert text at a specified index
+
+         "Hello World!".insert 5 " Cruel" == "Hello Cruel World!"
+         "Hello World!".insert -1 " Cruel" == "Hello World! Cruel"
+Text.insert : Integer -> Text -> Text ! Index_Out_Of_Bounds_Error
+Text.insert index that =
+    len = this.length
+    idx = if index < 0 then len + index + 1 else index
+    if (idx < 0) || (idx > len) then Error.throw (Index_Out_Of_Bounds_Error index len) else
+        if idx == 0 then Text_Utils.concat that this else
+            if idx == len then Text_Utils.concat this that else
+                pre = Text_Utils.substring this 0 idx
+                post = Text_Utils.substring this idx len
```

Comment:
I think `Text_Utils.substring` is a lower-level method which will not handle boundaries between grapheme clusters well.

See for example texts like `t = 'a\u{301}\u0915\u094D\u0937\u093F\u{1F926}\u{1F3FC}\u200D\u2642\uFE0F'` which will render to "áक्षि🤦🏼‍♂️" and so have length 3.
In such case, I think given that you use the grapheme-cluster-based-length (which will return 3), but use the raw substring, `t.insert 0 "foo"` will likely yield a text which does not include the emoji at the end.

Reply:
You can probably re-use `take`, i.e. `this.take (Range 0 idx)` which will handle finding the actual grapheme clusters.